mempool/clist_mempool.go

```diff
@@ -664,10 +712,136 @@ func (mem *CListMempool) recheckTxs() {
+	waitingForReap bool
+	reapSignal     chan struct{}
+	// This function is expected to run under recheckReapSharedState lock
+	// and can therefore safely access variables in here.
+	readyForReap func() bool
+}
+
+func (r *recheckReapSharedState) recheckCompleted() {
+	r.mtx.Lock()
+	r.successfullyUpdatedTxs = 0
+	r.bytesUpdated = 0
+	r.gasUpdated = 0
+	if r.waitingForReap {
+		r.sendReapSignal()
+	}
+	r.readyForReap = nil
+	r.reapSignal = nil
+	r.mtx.Unlock()
+}
+
+func (r *recheckReapSharedState) reapingCompleted() {
+	r.mtx.Lock()
+	r.isReaping = false
+	r.mtx.Unlock()
+}
+
+const updateBlockedOnReapPollInterval = 5 * time.Millisecond
+
+// This method protects against an edge case that seems impossible in practice.
+// We are still reaping, while we are committing the next block.
+// However reaping only occurs during propose, which is under the consensus mutex.
+// Update only happens during Commit which is also under the consensus mutex.
+// Furthermore Reap is very quick once unlocked.
+// If we are in this edge case that seems impossible, we just retry in a loop until recheck is complete.
+// Unclear if this method is needed in the first place.
+func (r *recheckReapSharedState) waitForReapingToFinish() {
+	r.mtx.Lock()
+	isReaping := r.isReaping
+	r.mtx.Unlock()
+	if !isReaping {
+		return
+	}
+	for {
+		time.Sleep(updateBlockedOnReapPollInterval)
+		r.mtx.Lock()
+		isReaping := r.isReaping
+		r.mtx.Unlock()
+		if !isReaping {
+			return
+		}
+	}
+}
+
+var ErrReapAlreadyWaiting = errors.New("reap already waiting. Only one reap can be waiting for Update to terminate at once")
+
+// TODO: Consider relaxing this, allow multiple concurrent reaps.
+// Not done to save on testing complexity, but should be a minimal code change if the feature is needed.
+var ErrAlreadyReaping = errors.New("already doing another reap instance. Only one reap can occur at once")
+
+func (r *recheckReapSharedState) setupWaitingReap(readyForReap func() bool) (chan struct{}, error) {
+	r.mtx.Lock()
+	if r.waitingForReap {
+		r.mtx.Unlock()
+		return nil, ErrReapAlreadyWaiting
+	}
+	if r.isReaping {
+		r.mtx.Unlock()
+		return nil, ErrAlreadyReaping
+	}
+	r.waitingForReap = true
+	r.readyForReap = readyForReap
+	newChan := make(chan struct{}, 1)
+	r.reapSignal = newChan
+	r.checkAndSendReapSignal()
+	r.mtx.Unlock()
+	return newChan, nil
+}
+
+// this function assumes we are under lock.
+func (r *recheckReapSharedState) checkAndSendReapSignal() {
+	if r.waitingForReap && r.readyForReap() {
+		r.sendReapSignal()
+	}
+}
+
+func (r *recheckReapSharedState) sendReapSignal() {
+	r.waitingForReap = false
+	r.reapSignal <- struct{}{}
+	r.reapSignal = nil
+	r.isReaping = true
+}
+
+func (rc *recheck) updateForRecheckSuccess(tx types.Tx, res *abci.CheckTxResponse) {
+	if rc.initialMempoolSize > minimumMempoolSizeForConcurrentRecheck {
+		rc.recheckReapSharedState.mtx.Lock()
+		rc.recheckReapSharedState.successfullyUpdatedTxs++
+		// TODO: compute proto size for txs may be slow.
```

Comment:
This is super inefficient and of course the tx size should be part of the tx entry on the mempool, computed once it is accepted by the application at first.

But this is a problem for another issue/PR.